business/openshift_oauth.go

```diff
@@ -0,0 +1,205 @@
+	TLS  *OAuthRouteTLSSpec `json:"tls,omitempty"`
+}
+
+type OAuthRouteTLSSpec struct {
+	Termination string `json:"termination"`
+}
+
+const serverPrefix = "https://kubernetes.default.svc/"
+
+func (in *OpenshiftOAuthService) Metadata() (metadata *OAuthMetadata, err error) {
+	var server *OAuthAuthorizationServer
+
+	response, err := request("GET", ".well-known/oauth-authorization-server", nil)
+
+	if err != nil {
+		message := fmt.Errorf("Could not send request to the Openshift OAuth API: %v", err)
+		return nil, message
+	}
+
+	err = json.Unmarshal(response, &server)
+
+	if err != nil {
+		message := fmt.Errorf("Could not parse data from the Openshift API: %v", err)
+		return nil, message
+	}
+
+	redirectURL, err := getKialiRoutePath()
+
+	if err != nil {
+		message := fmt.Errorf("Could not get Kiali route for OAuth redirect: %v", err)
+		return nil, message
+	}
+
+	metadata = &OAuthMetadata{}
+
+	metadata.AuthorizationEndpoint = fmt.Sprintf("%s?client_id=%s&redirect_uri=%s&response_type=%s", server.AuthorizationEndpoint, "kiali", url.QueryEscape(*redirectURL), "token")
+
+	return metadata, nil
+}
+
+func (in *OpenshiftOAuthService) ValidateToken(token string) error {
+	k8sConfig, err := kubernetes.ConfigClient()
+
+	if err != nil {
+		return fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	k8sConfig.BearerToken = token
+
+	k8s, err := kube.NewForConfig(k8sConfig)
+
+	if err != nil {
+		return fmt.Errorf("could not get Openshift cluster config: %v", err)
+	}
+
+	_, err = k8s.Discovery().ServerVersion()
+
+	if err != nil {
+		return fmt.Errorf("could not get info from Openshift: %v", err)
+	}
+
+	return nil
+}
+
+func (in *OpenshiftOAuthService) GetUserInfo(token string) (*OAuthUser, error) {
+	var user *OAuthUser
+
+	response, err := request("GET", "oapi/v1/users/~", &token)
+
+	if err != nil {
+		return nil, fmt.Errorf("Could not get user info from Openshift: %v", err)
+	}
+
+	err = json.Unmarshal(response, &user)
+
+	if err != nil {
+		return nil, fmt.Errorf("Could not parse user info from Openshift: %v", err)
+	}
+
+	return user, nil
+}
+
+func getKialiRoutePath() (*string, error) {
+	var route *OAuthRoute
+	var protocol string
+
+	namespace := config.Get().IstioNamespace
+
+	conf, err := kubernetes.ConfigClient()
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	response, err := request("GET", fmt.Sprintf("oapi/v1/namespaces/%s/routes/kiali", namespace), &conf.BearerToken)
+
+	err = json.Unmarshal(response, &route)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to Openshift: %v", err)
+	}
+
+	if route.Spec.TLS == nil {
+		protocol = "http://"
+	} else {
+		protocol = "https://"
+	}
+
+	url := strings.Join([]string{protocol, route.Spec.Host}, "")
+
+	return &url, nil
+}
+
+func request(method string, url string, auth *string) ([]byte, error) {
+	certPool := x509.NewCertPool()
+	cert, err := ioutil.ReadFile("/run/secrets/kubernetes.io/serviceaccount/ca.crt")
```

Comment:
is this always in the same spot? Or should we make this configurable in case the user mounts it in a different place?

Reply:
I don't think this is configurable. Its set by the kubernetes system itself.